pczt/src/roles/signer/mod.rs

```diff
@@ -0,0 +1,311 @@
+            .ok_or(Error::InvalidIndex)?;
+
+        // Check consistency of the input being signed.
+        // TODO
+
+        input
+            .sign(index, &self.tx_data, &self.txid_parts, sk, &self.secp)
+            .map_err(Error::TransparentSign)?;
+
+        // Update transaction modifiability:
+        // - If the Signer added a signature that does not use SIGHASH_ANYONECANPAY, the
+        //   Input Modifiable flag must be set to False.
+        if input.sighash_type().encode() & SIGHASH_ANYONECANPAY == 0 {
+            self.global.tx_modifiable &= !FLAG_INPUTS_MODIFIABLE;
+        }
+        // - If the Signer added a signature that does not use SIGHASH_NONE, the Outputs
+        //   Modifiable flag must be set to False.
+        if (input.sighash_type().encode() & !SIGHASH_ANYONECANPAY) != SIGHASH_NONE {
+            self.global.tx_modifiable &= !FLAG_OUTPUTS_MODIFIABLE;
+        }
+        // - If the Signer added a signature that uses SIGHASH_SINGLE, the Has SIGHASH_SINGLE
+        //   flag must be set to True.
+        if (input.sighash_type().encode() & !SIGHASH_ANYONECANPAY) == SIGHASH_SINGLE {
+            self.global.tx_modifiable |= FLAG_HAS_SIGHASH_SINGLE;
+        }
+
+        Ok(())
+    }
+
+    /// Signs the Sapling spend at the given index with the given spend authorizing key.
+    ///
+    /// It is the caller's responsibility to perform any semantic validity checks on the
+    /// PCZT (for example, comfirming that the change amounts are correct) before calling
+    /// this method.
+    pub fn sign_sapling(
+        &mut self,
+        index: usize,
+        ask: &sapling::keys::SpendAuthorizingKey,
+    ) -> Result<(), Error> {
+        let spend = self
+            .sapling
+            .spends_mut()
+            .get_mut(index)
+            .ok_or(Error::InvalidIndex)?;
+
+        // Check consistency of the input being signed.
+        let note_from_fields = spend
+            .recipient()
+            .zip(spend.value().as_ref())
+            .zip(spend.rseed().as_ref())
+            .map(|((recipient, value), rseed)| {
+                sapling::Note::from_parts(recipient, *value, *rseed)
+            });
+
+        if let Some(note) = note_from_fields {
+            let tx_spend = self
+                .tx_data
+                .sapling_bundle()
+                .expect("index checked above")
+                .shielded_spends()
+                .get(index)
+                .expect("index checked above");
+
+            let proof_generation_key = spend
+                .proof_generation_key()
+                .as_ref()
+                .ok_or(Error::MissingProofGenerationKey)?;
+
+            let nk = proof_generation_key.to_viewing_key().nk;
+
+            let merkle_path = spend.witness().as_ref().ok_or(Error::MissingWitness)?;
+
+            if &note.nf(&nk, merkle_path.position().into()) != tx_spend.nullifier() {
+                return Err(Error::InvalidNullifier);
+            }
+        }
+
```

Comment:
```suggestion

        // The sighash for a Sapling spend depends on all effecting data.
        self.tx_modifiable &= !(FLAG_INPUTS_MODIFIABLE | FLAG_OUTPUTS_MODIFIABLE);

```
(If `spend.sign` is guaranteed to be non-mutating when it fails, this could be after the call.)

Reply:
I originally didn't include these because I believed that shielded signatures couldn't be applied before the IO Finalizer ran, and that sets the necessary flags. However, looking at it again I think that these could potentially be reordered, so I'll ensure the flags get set regardless for shielded signatures.